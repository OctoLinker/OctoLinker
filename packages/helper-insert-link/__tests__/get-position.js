import getPosition from '../get-position';

describe('get-position', () => {
  const DEFAULT_REGEX = /(hello)/g;
  const REGEX_MULTI_LINES = /START\n(hello)\nEND/g;
  const REGEX_QUOTES = /(["']hello["'])/g;

  test.each([
    ['hello', DEFAULT_REGEX, 'hello'],
    ['foo hello', DEFAULT_REGEX, 'hello'],
    ['foo hello bar', DEFAULT_REGEX, 'hello'],
    ['foo\nhello bar', DEFAULT_REGEX, 'hello'],
    ['START\nhello\nEND', REGEX_MULTI_LINES, 'hello'],
    ['foo\nSTART\nhello\nEND', REGEX_MULTI_LINES, 'hello'],
    ['foo\nSTART\nhello\nEND', REGEX_MULTI_LINES, 'hello'],
    ['hello "hello"', REGEX_QUOTES, '"hello"'],
    [`hello 'hello'`, REGEX_QUOTES, "'hello'"],
    ['foo hello "hello"', REGEX_QUOTES, '"hello"'],
    ['foo hello "hello" bar', REGEX_QUOTES, '"hello"'],
    ['foo\nhello "hello" bar', REGEX_QUOTES, '"hello"'],
  ])('match input "%s" with regexp "%s"', (input, regex, expectedOutput) => {
    const result = getPosition(input, regex);
    const [{ lineNumber, startPos, endPos }] = result;

    expect(
      input.split('\n')[lineNumber - 1].substring(startPos, endPos),
    ).toEqual(expectedOutput);
    expect(result).toMatchSnapshot();
  });

  test('returns an empty array when regex does not match anything', () => {
    expect(getPosition('nomatch', DEFAULT_REGEX)).toStrictEqual([]);
  });

  test('returns an empty array when capture group is empty', () => {
    expect(getPosition('foo:bar', /foo:([0-9])?/g)).toStrictEqual([]);
  });
  test('returns an empty array when capture group is empty', () => {
    expect(getPosition('foo:bar', /(foo):(bar)/g)).toMatchSnapshot();
  });
});
