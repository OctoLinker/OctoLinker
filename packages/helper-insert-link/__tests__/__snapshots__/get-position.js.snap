// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-position match input "START
hello
END" with regexp "/START\\n(hello)\\nEND/g" 1`] = `
Array [
  Object {
    "endPos": 5,
    "endPosInBlob": 11,
    "lineNumber": 2,
    "startPos": 0,
    "startPosInBlob": 6,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo
START
hello
END" with regexp "/START\\n(hello)\\nEND/g" 1`] = `
Array [
  Object {
    "endPos": 5,
    "endPosInBlob": 15,
    "lineNumber": 3,
    "startPos": 0,
    "startPosInBlob": 10,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo
START
hello
END" with regexp "/START\\n(hello)\\nEND/g" 2`] = `
Array [
  Object {
    "endPos": 5,
    "endPosInBlob": 15,
    "lineNumber": 3,
    "startPos": 0,
    "startPosInBlob": 10,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo
hello "hello" bar" with regexp "/(["']hello["'])/g" 1`] = `
Array [
  Object {
    "endPos": 13,
    "endPosInBlob": 17,
    "lineNumber": 2,
    "startPos": 6,
    "startPosInBlob": 10,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo
hello bar" with regexp "/(hello)/g" 1`] = `
Array [
  Object {
    "endPos": 5,
    "endPosInBlob": 9,
    "lineNumber": 2,
    "startPos": 0,
    "startPosInBlob": 4,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo hello "hello" bar" with regexp "/(["']hello["'])/g" 1`] = `
Array [
  Object {
    "endPos": 17,
    "endPosInBlob": 17,
    "lineNumber": 1,
    "startPos": 10,
    "startPosInBlob": 10,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo hello "hello"" with regexp "/(["']hello["'])/g" 1`] = `
Array [
  Object {
    "endPos": 17,
    "endPosInBlob": 17,
    "lineNumber": 1,
    "startPos": 10,
    "startPosInBlob": 10,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo hello bar" with regexp "/(hello)/g" 1`] = `
Array [
  Object {
    "endPos": 9,
    "endPosInBlob": 9,
    "lineNumber": 1,
    "startPos": 4,
    "startPosInBlob": 4,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "foo hello" with regexp "/(hello)/g" 1`] = `
Array [
  Object {
    "endPos": 9,
    "endPosInBlob": 9,
    "lineNumber": 1,
    "startPos": 4,
    "startPosInBlob": 4,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "hello "hello"" with regexp "/(["']hello["'])/g" 1`] = `
Array [
  Object {
    "endPos": 13,
    "endPosInBlob": 13,
    "lineNumber": 1,
    "startPos": 6,
    "startPosInBlob": 6,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "hello 'hello'" with regexp "/(["']hello["'])/g" 1`] = `
Array [
  Object {
    "endPos": 13,
    "endPosInBlob": 13,
    "lineNumber": 1,
    "startPos": 6,
    "startPosInBlob": 6,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position match input "hello" with regexp "/(hello)/g" 1`] = `
Array [
  Object {
    "endPos": 5,
    "endPosInBlob": 5,
    "lineNumber": 1,
    "startPos": 0,
    "startPosInBlob": 0,
    "values": Array [
      "hello",
    ],
  },
]
`;

exports[`get-position returns an empty array when capture group is empty 1`] = `
Array [
  Object {
    "endPos": 3,
    "endPosInBlob": 3,
    "lineNumber": 1,
    "startPos": 0,
    "startPosInBlob": 0,
    "values": Array [
      "foo",
      "bar",
    ],
  },
]
`;
